File: android/src/main/kotlin/com/olo/flutter/olo_pay_sdk/controls/singleline/PaymentCardDetailsSingleLineView.kt
// Copyright © 2022 Olo Inc. All rights reserved.
// This software is made available under the Olo Pay SDK License (See LICENSE.md file)
package com.olo.flutter.olo_pay_sdk.controls.singleline

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Configuration
import android.graphics.Typeface
import android.os.Build
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import com.olo.flutter.olo_pay_sdk.R
import com.olo.flutter.olo_pay_sdk.data.CustomErrorMessages
import com.olo.flutter.olo_pay_sdk.data.DataKeys
import com.olo.flutter.olo_pay_sdk.data.ErrorCodes
import com.olo.flutter.olo_pay_sdk.data.GlobalConstants
import com.olo.flutter.olo_pay_sdk.extensions.getArgOrErrorResult
import com.olo.flutter.olo_pay_sdk.extensions.toMap
import com.olo.flutter.olo_pay_sdk.utils.OloPayLog
import com.olo.olopay.api.OloPayAPI
import com.olo.olopay.controls.callbacks.CardInputListener
import com.olo.olopay.controls.callbacks.ConfigurationChangeListener
import com.olo.olopay.data.CardErrorType
import com.olo.olopay.data.CardField
import com.olo.olopay.data.ICardFieldState
import com.olo.olopay.exceptions.ApiConnectionException
import com.olo.olopay.exceptions.ApiException
import com.olo.olopay.exceptions.CardException
import com.olo.olopay.exceptions.InvalidRequestException
import com.olo.olopay.exceptions.OloPayException
import com.olo.olopay.exceptions.RateLimitException
import com.stripe.android.core.exception.AuthenticationException
import io.flutter.FlutterInjector
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import com.olo.olopay.controls.PaymentCardDetailsSingleLineView as OloPaySingleLineView


@SuppressLint("InflateParams")
internal class PaymentCardDetailsSingleLineView(
    context: Context,
    messenger: BinaryMessenger,
    id: Int,
    args: Any?
) : PlatformView, MethodChannel.MethodCallHandler, CardInputListener, ConfigurationChangeListener {
    private val methodChannel: MethodChannel

    private var cardInputView: OloPaySingleLineView = LayoutInflater.from(context).inflate(
        R.layout.flutter_olopay_single_line_view,
        null
    ) as OloPaySingleLineView


    private val defaultFont = cardInputView.getFont()
    private var _customErrorMessages: CustomErrorMessages? = null

    init {
        cardInputView.configurationChangeListener = this
        cardInputView.displayErrors = false
        cardInputView.cardInputListener = this

        loadCustomArgs(args)

        methodChannel = MethodChannel(messenger, DataKeys.SingleLineBaseMethodChannelKey + id)
        methodChannel.setMethodCallHandler(this)
    }

    override fun getView(): View {
        return cardInputView
    }

    override fun dispose(){}

    // Device orientation change apparently does not follow the same flow as in native Android, which
    // results in the layout not getting updated properly for new widths. The size gets updated, but
    // each individual field size is not, resulting in overlapping fields or inaccessible fields that
    // are outside the bounds of the view. This little hack fixes that. By forcing focus on specific
    // fields, it triggers a recalculation of Stripe's individual fields, and fixes this issue.
    // As of right now, this appears to be the only way to fix this issue, as traditional methods
    // involving invalidate(), requestLayout(), and forceLayout() do not work.
    override fun onConfigurationChanged(newConfig: Configuration?) {
        val initialFocused: CardField? = getFocusedField(cardInputView.fieldStates)

        // Switching focus to these fields guarantees a recalculation of field size on Stripe's
        // control will occur.
        val newFocusedField = when (initialFocused) {
            CardField.CardNumber -> CardField.Expiration
            CardField.Expiration -> CardField.CardNumber
            CardField.Cvv -> CardField.CardNumber
            CardField.PostalCode -> CardField.CardNumber
            null -> CardField.Expiration
        }

        // If there wasn't a focused field, setting the focus on the CardNumber field makes the most
        // sense since it's the first field users interact with, and this second focus is the key
        // to getting the control to display properly in the new configuration.
        //
        // Attempts to clear focus after this, in the case that a field didn't have focus to begin
        // with, appear to cause too much work on the UI thread, and focus doesn't consistently get
        // cleared. Rather than have inconsistent results, we just don't attempt to clear the focus.
        val restoredFocusField = initialFocused ?: CardField.CardNumber

        // Important: This should not be in a post call
        cardInputView.requestFocus(newFocusedField, false)

        // This delay is important, as it allows the view time to finish the previous focus call
        // before attempting this one. The native Android SDK has logic where, in certain scenarios,
        // It can take up to 300ms to request focus, so this delay should never be less than that.
        cardInputView.postDelayed({
            cardInputView.requestFocus(restoredFocusField, false)
        }, FocusDelayMs)
    }

    override fun onMethodCall(methodCall: MethodCall, result: MethodChannel.Result) {
        when(methodCall.method) {
            DataKeys.CreatePaymentMethodKey -> createPaymentMethod(result)
            DataKeys.GetStateMethodKey -> getState(result)
            DataKeys.IsValidMethodKey -> isValid(result)
            DataKeys.GetCardTypeMethodKey -> getCardType(result)
            DataKeys.IsEnabledMethodKey -> isEnabled(result)
            DataKeys.HasErrorMessageMethodKey -> hasErrorMessage(methodCall, result)
            DataKeys.GetErrorMessageMethodKey -> getErrorMessage(methodCall, result)
            DataKeys.ClearFieldsMethodKey -> clearFields(result)
            DataKeys.RequestFocusMethodKey -> requestFocus(result)
            DataKeys.ClearFocusMethodKey -> clearFocus(result)
            DataKeys.RefreshUiMethod -> refreshUI(methodCall, result)
            else -> result.notImplemented()
        }
    }

    private fun createPaymentMethod(result: MethodChannel.Result) {
        if (!cardInputView.isValid) {
            val errorCode = getErrorCode()
            val errorMessage = getErrorMessage(false)
            result.error(errorCode, errorMessage, null)
            return
        }

        CoroutineScope(Dispatchers.IO).launch {
            val params = cardInputView.paymentMethodParams!!

            try {
                val paymentMethod = OloPayAPI().createPaymentMethod(cardInputView.context, params)
                result.success(paymentMethod.toMap())
            } catch (e: ApiException) {
                result.error(ErrorCodes.ApiError, e.message, null)
            } catch(e: InvalidRequestException) {
                result.error(ErrorCodes.InvalidRequest, e.message, null)
            } catch(e: ApiConnectionException) {
                result.error(ErrorCodes.Connection, e.message, null)
            } catch(e: RateLimitException) {
                result.error(ErrorCodes.RateLimit, e.message, null)
            } catch(e: AuthenticationException) {
                result.error(ErrorCodes.Authentication, e.message, null)
            } catch(e: CardException) {
                when(e.type) {
                    CardErrorType.InvalidNumber -> result.error(ErrorCodes.InvalidNumber, e.message, null)
                    CardErrorType.InvalidExpMonth -> result.error(ErrorCodes.InvalidExpiration, e.message, null)
                    CardErrorType.InvalidExpYear -> result.error(ErrorCodes.InvalidExpiration, e.message, null)
                    CardErrorType.InvalidCVV -> result.error(ErrorCodes.InvalidCvv, e.message, null)
                    CardErrorType.InvalidZip -> result.error(ErrorCodes.InvalidPostalCode, e.message, null)
                    CardErrorType.ExpiredCard -> result.error(ErrorCodes.ExpiredCard, e.message, null)
                    CardErrorType.CardDeclined -> result.error(ErrorCodes.CardDeclined, e.message, null)
                    CardErrorType.ProcessingError -> result.error(ErrorCodes.ProcessingError, e.message, null)
                    CardErrorType.UnknownCardError -> result.error(ErrorCodes.UnknownCardError, e.message, null)
                }
            } catch(e: OloPayException) {
                result.error(ErrorCodes.GeneralError, e.message, null)
            }
        }
    }

    private fun getState(result: MethodChannel.Result) {
        result.success(cardInputView.fieldStates.toMap())
    }

    private fun isValid(result: MethodChannel.Result) {
        result.success(cardInputView.isValid)
    }

    private fun getCardType(result: MethodChannel.Result) {
        result.success(cardInputView.cardBrand.description)
    }

    private fun isEnabled(result: MethodChannel.Result) {
        result.success(cardInputView.isEnabled)
    }

    private fun hasErrorMessage(call: MethodCall, result: MethodChannel.Result) {
        val ignoreUneditedFields = try {
            call.getArgOrErrorResult<Boolean>(
                DataKeys.IgnoreUneditedFieldsParameterKey,
                "Unable to check for error message",
                result
            )
        } catch (_: Exception) {
            return
        }

        result.success(cardInputView.hasErrorMessage(ignoreUneditedFields))
    }

    private fun getErrorMessage(call: MethodCall, result: MethodChannel.Result) {
        val ignoreUneditedFields = try {
            call.getArgOrErrorResult<Boolean>(
                DataKeys.IgnoreUneditedFieldsParameterKey,
                "Unable to get error message",
                result
            )
        } catch (_: Exception) {
            return
        }

        result.success(getErrorMessage(ignoreUneditedFields))
    }

    private fun getErrorMessage(ignoreUneditedFields: Boolean): String {
        if(cardInputView.isValid || !cardInputView.hasErrorMessage(ignoreUneditedFields)) {
            return ""
        }

        val defaultErrorMessage = cardInputView.getErrorMessage(ignoreUneditedFields)
        return (_customErrorMessages?.getCustomErrorMessage(ignoreUneditedFields, cardInputView.fieldStates, cardInputView.cardBrand) ?: defaultErrorMessage)
    }

    private fun clearFields(result: MethodChannel.Result) {
        cardInputView.clearFields()
        result.success(null)
    }

    private fun requestFocus(result: MethodChannel.Result) {
        cardInputView.requestFocus(CardField.CardNumber, true)
        result.success(null)
    }

    private fun clearFocus(result: MethodChannel.Result) {
        cardInputView.clearFocus()
        result.success(null)
    }

    private fun refreshUI(call: MethodCall, result: MethodChannel.Result) {
        call.argument<Any>(DataKeys.CreationParameters)?.let { params ->
            loadCustomArgs(params)
        }
        result.success(null)
    }

    private fun getErrorCode(): String {
        if(!cardInputView.fieldStates[CardField.CardNumber]!!.isValid) {
            return ErrorCodes.InvalidNumber
        } else if (!cardInputView.fieldStates[CardField.Expiration]!!.isValid) {
            return ErrorCodes.InvalidExpiration
        } else if(!cardInputView.fieldStates[CardField.Cvv]!!.isValid) {
            return ErrorCodes.InvalidCvv
        } else if(!cardInputView.fieldStates[CardField.PostalCode]!!.isValid) {
            return ErrorCodes.InvalidPostalCode
        }

        return ErrorCodes.InvalidCardDetails
    }

    override fun onFocusChange(field: CardField?, fieldStates: Map<CardField, ICardFieldState>) {
        val args = mapOf(
            DataKeys.IsValidKey to cardInputView.isValid,
            DataKeys.FocusedFieldParameterKey to (field?.toString() ?: ""),
            DataKeys.FieldStatesParameterKey to fieldStates.toMap()
        )

        methodChannel.invokeMethod(DataKeys.OnFocusChangedEventHandlerKey, args)
        emitErrorMessage()
    }

    override fun onValidStateChanged(isValid: Boolean, fieldStates: Map<CardField, ICardFieldState>) {
        val args = mapOf(
            DataKeys.IsValidKey to isValid,
            DataKeys.FieldStatesParameterKey to fieldStates.toMap()
        )

        methodChannel.invokeMethod(DataKeys.OnValidStateChangedEventHandlerKey, args)
        emitErrorMessage()
    }

    override fun onInputChanged(isValid: Boolean, fieldStates: Map<CardField, ICardFieldState>) {
        val args = mapOf(
            DataKeys.IsValidKey to isValid,
            DataKeys.FieldStatesParameterKey to fieldStates.toMap()
        )

        methodChannel.invokeMethod(DataKeys.OnInputChangedEventHandlerKey, args)
        emitErrorMessage()
    }

    private fun emitErrorMessage() {
        methodChannel.invokeMethod(DataKeys.OnErrorMessageChangedEventHandlerKey, getErrorMessage(true))
    }

    private fun loadCustomArgs(args: Any?) {
        val widgetArgs = args as? Map<*, *> ?: return

        val hints = widgetArgs[DataKeys.HintsArgumentKey] as? Map<*, *>
        if (hints != null && hints.isNotEmpty()) {
            loadHints(hints)
        }

        val textStyles = widgetArgs[DataKeys.TextStylesArgumentsKey] as? Map<*, *>
        if (textStyles != null && textStyles.isNotEmpty()) {
            loadTextStyles(textStyles)
        }

        val backgroundStyles = widgetArgs[DataKeys.BackgroundStylesArgumentsKey] as? Map<*, *>
        if (backgroundStyles != null && backgroundStyles.isNotEmpty()) {
            loadBackgroundStyles(backgroundStyles)
        }

        val paddingStyles = widgetArgs[DataKeys.PaddingStylesArgumentsKey] as? Map<*, *>
        if (paddingStyles != null && paddingStyles.isNotEmpty()) {
            loadPaddingStyles(paddingStyles)
        }

        val customErrorMessages = widgetArgs[DataKeys.CustomErrorMessagesArgumentsKey] as? Map<*, *>
        _customErrorMessages = if (!customErrorMessages.isNullOrEmpty()) {
            CustomErrorMessages(customErrorMessages)
        } else {
            null
        }

        (widgetArgs[DataKeys.EnabledParameterKey] as? Boolean)?.let {
            cardInputView.isEnabled = it
        }
    }

    private fun loadHints(hints: Map<*, *>) {
        val numberHint = hints[CardField.CardNumber.toString()] as? String
        if (numberHint != null) {
            cardInputView.setHintText(CardField.CardNumber, numberHint)
        }

        val expirationHint = hints[CardField.Expiration.toString()] as? String
        if (expirationHint != null) {
            cardInputView.setHintText(CardField.Expiration, expirationHint)
        }

        val cvvHint = hints[CardField.Cvv.toString()] as? String
        if (cvvHint != null) {
            cardInputView.setHintText(CardField.Cvv, cvvHint)
        }

        val postalCodeHint = hints[CardField.PostalCode.toString()] as? String
        if (postalCodeHint != null) {
            cardInputView.setHintText(CardField.PostalCode, postalCodeHint)
        }
    }

    private fun loadTextStyles(textStyles: Map<*, *>) {
        if(Build.VERSION.SDK_INT >= GlobalConstants.ApiOreo) {
            val textColor = textStyles[DataKeys.TextColorKey] as? String
            if (!textColor.isNullOrBlank()) {
                cardInputView.setTextColor(textColor)
            }

            val errorTextColor = textStyles[DataKeys.ErrorTextColorKey] as? String
            if (!errorTextColor.isNullOrBlank()) {
                cardInputView.setErrorTextColor(errorTextColor)
            }

            val hintTextColor = textStyles[DataKeys.HintTextColorKey] as? String
            if (!hintTextColor.isNullOrBlank()) {
                cardInputView.setHintTextColor(hintTextColor)
            }
        }

        val cursorColor = textStyles[DataKeys.CursorColorKey] as? String
        if (!cursorColor.isNullOrBlank() && Build.VERSION.SDK_INT >= GlobalConstants.ApiQuinceTart) {
            cardInputView.setCursorColor(cursorColor)
        }

        (textStyles[DataKeys.TextSizeKey] as? Double)?.let {
            cardInputView.setTextSize(it.toFloat())
        }

        val fontAsset = textStyles[DataKeys.FontAssetKey] as? String
        if(!fontAsset.isNullOrBlank()) {
            try{
                val assetKey = FlutterInjector.instance().flutterLoader().getLookupKeyForAsset(fontAsset)
                val font = Typeface.createFromAsset(cardInputView.context.assets, assetKey)
                cardInputView.setFont(font)
            } catch (error: Exception) {
                OloPayLog.e(OloPayLog.getStackTrace(error))
                cardInputView.setFont(defaultFont)
            }
        } else if(cardInputView.getFont() != defaultFont){
            cardInputView.setFont(defaultFont)
        }
    }

    private fun loadBackgroundStyles(backgroundStyles: Map<*, *>) {
        val displayMetrics = cardInputView.context.resources.displayMetrics

        //Need to convert from dp values to px values
        val borderWidthPx = (backgroundStyles[DataKeys.BorderWidthKey] as? Double)?.let {
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, it.toFloat(), displayMetrics)
        }

        val borderRadiusPx = (backgroundStyles[DataKeys.BorderRadiusKey] as? Double)?.let {
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, it.toFloat(), displayMetrics)
        }

        if (Build.VERSION.SDK_INT >= GlobalConstants.ApiOreo) {
            cardInputView.setCardBackgroundStyle(
                backgroundColorHex = backgroundStyles[DataKeys.BackgroundColorKey] as? String,
                borderColorHex = backgroundStyles[DataKeys.BorderColorKey] as? String,
                borderWidthPx = borderWidthPx,
                borderRadiusPx = borderRadiusPx)
        }
    }

    private fun loadPaddingStyles(paddingStyles: Map<*, *>) {
        val displayMetrics = cardInputView.context.resources.displayMetrics

        val startPadding = (paddingStyles[DataKeys.StartPaddingKey] as? Double)?.let {
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, it.toFloat(), displayMetrics).toInt()
        }

        val endPadding = (paddingStyles[DataKeys.EndPaddingKey] as? Double)?.let {
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, it.toFloat(), displayMetrics).toInt()
        }
        val topPadding = (paddingStyles[DataKeys.TopPaddingKey] as? Double)?.let {
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, it.toFloat(), displayMetrics).toInt()
        }

        val bottomPadding = (paddingStyles[DataKeys.BottomPaddingKey] as? Double)?.let {
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, it.toFloat(), displayMetrics).toInt()
        }

        cardInputView.setCardPadding(
            startPx = startPadding,
            endPx = endPadding,
            topPx = topPadding,
            bottomPx = bottomPadding
        )
    }

    private fun getFocusedField(state: Map<CardField, ICardFieldState>) : CardField? {
        state.forEach {
            if (it.value.isFocused) {
                return it.key
            }
        }

        return null
    }

    companion object {
        const val FocusDelayMs = 350L
    }
}

File: ios/Classes/PaymentCardDetailsSingleLineView/PaymentCardDetailsSingleLineView.swift
// Copyright © 2022 Olo Inc. All rights reserved.
// This software is made available under the Olo Pay SDK License (See LICENSE.md file)
//
//  PaymentCardDetailsSingleLineView.swift
//  olo_pay_sdk
//
//  Created by Justin Anderson on 12/7/23.
//

import Foundation
import Flutter
import UIKit
import OloPaySDK
import os.log

class PaymentCardDetailsSingleLineView: NSObject, FlutterPlatformView, OPPaymentCardDetailsViewDelegate {
    private let _logger = OSLog.init(subsystem: Bundle.main.bundleIdentifier!, category: DataKeys.BridgePrefix)
    
    private var _customErrorMessages: CustomErrorMessages? = nil
    private var _cardInputView: OPPaymentCardDetailsView
    private let _methodChannel: FlutterMethodChannel
    
    init(withFrame frame: CGRect, viewId: Int64, args: Any?, messenger: FlutterBinaryMessenger) {
        _cardInputView = OPPaymentCardDetailsView(frame: frame)
        _cardInputView.displayGeneratedErrorMessages = false
        
        let channelName = "\(DataKeys.SingleLineBaseMethodChannelKey)\(viewId)"
        _methodChannel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
        
        super.init()
        _methodChannel.setMethodCallHandler(onMethodCall(call:result:))
        
        _cardInputView.cardDetailsDelegate = self
        
        loadCustomArgs(args: args)
    }
    
    var customErrorMessages: Dictionary<String, Any>? = nil {
        didSet {
            if let customErrorMessages: Dictionary<String, Any> = customErrorMessages {
                _customErrorMessages = CustomErrorMessages(customErrorMessages: customErrorMessages as NSDictionary)
            } else {
                _customErrorMessages = nil
            }
        }
    }
    
    func view() -> UIView {
        return _cardInputView
    }
    
    private func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case DataKeys.CreatePaymentMethodKey:
            createPaymentMethod(result: result)
        case DataKeys.GetStateMethodKey:
            getState(result: result)
        case DataKeys.IsValidMethodKey:
            isValid(result: result)
        case DataKeys.GetCardTypeMethodKey:
            getCardType(result: result)
        case DataKeys.IsEnabledMethodKey:
            isEnabled(result: result)
        case DataKeys.HasErrorMessageMethodKey:
            hasErrorMessage(call: call, result: result)
        case DataKeys.GetErrorMessageMethodKey:
            getErrorMessage(call: call, result: result)
        case DataKeys.ClearFieldsMethodKey:
            clearFields(result: result)
        case DataKeys.RequestFocusMethodKey:
            requestFocus(call: call, result: result)
        case DataKeys.ClearFocusMethodKey:
            clearFocus(result: result)
        case DataKeys.RefreshUiMethodKey:
            refreshUI(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func createPaymentMethod(result: @escaping FlutterResult) {
        guard let params = _cardInputView.getPaymentMethodParams() else {
            result(FlutterError(code: getErrorCode(), message: getErrorMessage(false), details: nil))
            return
        }
        
        OloPayAPI().createPaymentMethod(with: params) { paymentMethod, error in
            if (error != nil) {
                rejectError(error: error!, result: result)
                return
            }

            guard let paymentMethod = paymentMethod else {
                result(FlutterError(code: ErrorCodes.GeneralError, message: "Unexpected error occurred", details: nil))
                return
            }

            result(paymentMethod.toDictionary())
        }
    }
    
    private func getState(result: @escaping FlutterResult) {
        result(_cardInputView.fieldStates.toDictionary())
    }
    
    private func isValid(result: @escaping FlutterResult) {
        result(_cardInputView.isValid)
    }
    
    private func getCardType(result: @escaping FlutterResult) {
        result(_cardInputView.cardType.description)
    }
    
    private func isEnabled(result: @escaping FlutterResult) {
        result(_cardInputView.isEnabled)
    }
    
    private func hasErrorMessage(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? Dictionary<String, Any>
        
        guard let ignoreUneditedFields: Bool = args?.get(DataKeys.IgnoreUneditedFieldsParameterKey) else {
            result(FlutterError(
                code: ErrorCodes.MissingParameter,
                message: "Missing Parameter \(DataKeys.IgnoreUneditedFieldsParameterKey)",
                details: nil)
            )
            return
        }
        
        result(_cardInputView.hasErrorMessage(ignoreUneditedFields))
    }
    
    private func getErrorMessage(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? Dictionary<String, Any>
        
        do {
            let ignoreUneditedFields: Bool = try args.getOrErrorResult(
                for: DataKeys.IgnoreUneditedFieldsParameterKey,
                baseError: "Unable to get error message",
                result: result
            )
            
            result(self.getErrorMessage(ignoreUneditedFields))
        } catch {
            return
        }
    }
    
    private func getErrorMessage(_ ignoreUneditedFieldErrors: Bool = true) -> String {
        if (_cardInputView.isValid || !_cardInputView.hasErrorMessage(ignoreUneditedFieldErrors)) {
            return ""
        }
        
        let defaultError = CustomErrorMessages.getDefaultErrorMessage(
            ignoreUneditedFieldErrors,
            _cardInputView.fieldStates,
            _cardInputView.cardType
        )
            
        var customError: String? = nil
        
        if let customErrorMessages = _customErrorMessages {
            customError = customErrorMessages.getCustomErrorMessage(
                ignoreUneditedFieldErrors,
                _cardInputView.fieldStates,
                _cardInputView.cardType
            )
        }
        
        return customError ?? defaultError
    }
    
    private func errorMessageHandler(_ cardState: NSDictionary, _ cardBrand: OPCardBrand, _ ignoreUneditedFieldErrors: Bool) -> String {
        return getErrorMessage(ignoreUneditedFieldErrors)
    }
    
    private func clearFields(result: @escaping FlutterResult) {
        _cardInputView.clear()
        result(nil)
    }
    
    private func requestFocus(call: FlutterMethodCall?, result: @escaping FlutterResult) {
        let args = call?.arguments as? Dictionary<String, Any>
        do {
            let focusField = try args?.getStringOrErrorResult(
                for: DataKeys.FocusFieldParameterKey,
                withDefault: DataKeys.CardNumberFieldValueKey,
                baseError: "Unable to set focus",
                acceptEmptyValue: false,
                result: result
                )
            _cardInputView.becomeFirstResponder(at: OPCardField.from(value: focusField ?? DataKeys.CardNumberFieldValueKey))
            result(nil)
        } catch {
            return
        }
    }
    
    private func clearFocus(result: @escaping FlutterResult) {
        _cardInputView.resignFirstResponder()
        result(nil)
    }
    
    private func refreshUI(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? Dictionary<String, Any>
        
        if let creationParams = args?.getDictionary(DataKeys.CreationParameters) {
            loadCustomArgs(args: creationParams)
        }
        
        result(nil)
    }
    
    private func getErrorCode() -> String {
        if !_cardInputView.fieldStates[OPCardField.number]!.isValid {
            return ErrorCodes.InvalidNumber
        } else if !_cardInputView.fieldStates[OPCardField.expiration]!.isValid {
            return ErrorCodes.InvalidExpiration
        } else if !_cardInputView.fieldStates[OPCardField.cvv]!.isValid {
            return ErrorCodes.InvalidCvv
        } else if !_cardInputView.fieldStates[OPCardField.postalCode]!.isValid {
            return ErrorCodes.InvalidPostalCode
        }
        
        return ErrorCodes.InvalidCardDetails
    }
    
    func paymentCardDetailsViewDidChange(with fieldStates: NSDictionary, isValid: Bool) {
        let args = [
            DataKeys.IsValidKey: isValid,
            DataKeys.FieldStatesParameterKey: _cardInputView.fieldStates.toDictionary()
        ] as [String : Any]
        
        _methodChannel.invokeMethod(DataKeys.OnInputChangedEventHandlerKey, arguments: args)
        emitErrorMessage()
    }
    
    func paymentCardDetailsViewFieldDidBeginEditing(with fieldStates: NSDictionary, field: OPCardField, isValid: Bool) {
        let args = [
            DataKeys.FocusedFieldParameterKey: field.flutterBridgeValue(),
            DataKeys.IsValidKey: isValid,
            DataKeys.FieldStatesParameterKey: _cardInputView.fieldStates.toDictionary()
        ] as [String : Any]
        
        _methodChannel.invokeMethod(DataKeys.OnFocusChangedEventHandlerKey, arguments: args)
        emitErrorMessage()
    }
    
    func paymentCardDetailsViewDidEndEditing(with fieldStates: NSDictionary, isValid: Bool) {
        let args = [
            DataKeys.FocusedFieldParameterKey: OPCardField.unknown.flutterBridgeValue(),
            DataKeys.IsValidKey: isValid,
            DataKeys.FieldStatesParameterKey: _cardInputView.fieldStates.toDictionary()
        ] as [String : Any]
        
        _methodChannel.invokeMethod(DataKeys.OnFocusChangedEventHandlerKey, arguments: args)
        emitErrorMessage()
    }
    
    func paymentCardDetailsViewIsValidChanged(with fieldStates: NSDictionary, isValid: Bool) {
        let args = [
            DataKeys.IsValidKey: isValid,
            DataKeys.FieldStatesParameterKey: _cardInputView.fieldStates.toDictionary()
        ] as [String : Any]
        
        _methodChannel.invokeMethod(DataKeys.OnValidStateChangedEventHandlerKey, arguments: args)
        emitErrorMessage()
    }
    
    private func emitErrorMessage() {
        _methodChannel.invokeMethod(
            DataKeys.OnErrorMessageChangedEventHandlerKey,
            arguments: getErrorMessage(true)
        )
    }
    
    private func loadCustomArgs(args: Any?) {
        guard let viewArgs = args as? Dictionary<String, Any> else {
            return
        }
        
        if let hints = viewArgs[DataKeys.HintsArgumentsKey] as? Dictionary<String, String?> {
            loadHints(hints: hints)
        }
        
        if let textStyles = viewArgs[DataKeys.TextStylesArgumentsKey] as? Dictionary<String, Any>, !textStyles.isEmpty {
            loadTextStyles(textStyles: textStyles)
        }
        
        if let backgroundStyles = viewArgs[DataKeys.BackgroundStylesArgumentsKey] as? Dictionary<String, Any>, !backgroundStyles.isEmpty {
            loadBackgroundStyles(backgroundStyles: backgroundStyles)
        }
        
        if let customErrorMessages = viewArgs[DataKeys.CustomErrorMessagesArgumentsKey] as? Dictionary<String, Any>, !customErrorMessages.isEmpty {
            self.customErrorMessages = customErrorMessages
        } else {
            self.customErrorMessages = nil
        }
        
        if let isEnabled = viewArgs[DataKeys.EnabledParameterKey] as? Bool {
            _cardInputView.isEnabled = isEnabled
        }
     
        OPPaymentCardDetailsView.errorMessageHandler = errorMessageHandler(_:_:_:)
    }
    
    private func loadHints(hints: Dictionary<String, String?>) {
        if let numberHint = hints[OPCardField.number.flutterBridgeValue()] {
            _cardInputView.numberPlaceholder = numberHint
        }
        
        if let expirationHint = hints[OPCardField.expiration.flutterBridgeValue()] {
            _cardInputView.expirationPlaceholder = expirationHint
        }
        
        if let cvvHint = hints[OPCardField.cvv.flutterBridgeValue()] {
            _cardInputView.cvvPlaceholder = cvvHint
        }
        
        if let postalCodeHint = hints[OPCardField.postalCode.flutterBridgeValue()] {
            _cardInputView.postalCodePlaceholder = postalCodeHint
        }
    }
    
    private func loadTextStyles(textStyles: Dictionary<String, Any>) {
        if let textColor = textStyles[DataKeys.TextColorKey] as? String, !textColor.isEmpty {
            _cardInputView.textColor = UIColor(hex: textColor)
        }
        
        if let errorTextColor = textStyles[DataKeys.ErrorTextColorKey] as? String, !errorTextColor.isEmpty {
            _cardInputView.textErrorColor = UIColor(hex: errorTextColor)
        }
        
        if let cursorColor = textStyles[DataKeys.CursorColorKey] as? String, !cursorColor.isEmpty {
            _cardInputView.cursorColor = UIColor(hex: cursorColor)
        }
        
        if let hintTextColor = textStyles[DataKeys.HintTextColorKey] as? String, !hintTextColor.isEmpty {
            _cardInputView.placeholderColor = UIColor(hex: hintTextColor)
        }
        
        if let fontSize = textStyles[DataKeys.TextSizeKey] as? Float {
            // We need a font size in order to set a font... thankfully we always send a font size from the
            // Flutter side of things
            let fontName = textStyles[DataKeys.FontNameKey] as? String
            
            var font = UIFont.systemFont(ofSize: CGFloat(fontSize))
            
            if let fontAsset = textStyles[DataKeys.FontAssetKey] as? String {
                do {
                    font = try UIFont.from(assetPath: fontAsset, with: fontName, size: fontSize)
                } catch OloError.AssetNotFoundError {
                    os_log("Font asset not found: %@", log: _logger, fontAsset)
                } catch OloError.FontLoadError {
                    os_log("Unable to load font: %@", log: _logger, fontAsset)
                } catch OloError.FontNameError {
                    os_log("Font name not found: %@", log: _logger, fontName ?? "")
                } catch {
                    os_log("Unexpected error loading font: %@", log: _logger, fontAsset)
                }
            }
            
            _cardInputView.font = UIFontMetrics.default.scaledFont(for: font)
        }
    }
    
    private func loadBackgroundStyles(backgroundStyles: Dictionary<String, Any>) {
        if let backgroundColor = backgroundStyles[DataKeys.BackgroundColorKey] as? String, !backgroundColor.isEmpty {
            _cardInputView.backgroundColor = UIColor(hex: backgroundColor)
        }
        
        if let borderColor = backgroundStyles[DataKeys.BorderColorKey] as? String, !borderColor.isEmpty {
            _cardInputView.borderColor = UIColor(hex: borderColor)
        }
        
        if let borderWidth = backgroundStyles[DataKeys.BorderWidthKey] as? Double {
            _cardInputView.borderWidth = CGFloat(borderWidth)
        }
        
        if let borderRadius = backgroundStyles[DataKeys.BorderRadiusKey] as? Double {
            _cardInputView.cornerRadius = CGFloat(borderRadius)
        }
    }
}
